overlord/devicestate/devicestate.go

```diff
@@ -641,6 +782,24 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
 		return nil
 	}
 
+	vSets, err := validationSetsFromModel(new, st, deviceCtx.Store(), len(localSnaps) > 0)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := vSets.Conflict(); err != nil {
+		return nil, err
+	}
+
+	if err := checkForInvalidSnapsInModel(new, vSets); err != nil {
+		return nil, err
+	}
+
+	// any snap that has a required revision will be in this map, if the snap's
+	// version is unconstrained, then we'll get a default-initialized revision
+	// from the map
+	snapRevisions := vSets.Revisions()
```

Comment:
This could be moved to a separate function instead of making `remodelTasks` bigger, it's already a pretty complex function. Something like `verifyAgainstModelValidationSets` or something else, that returns the snap revision map and an error if the validation failed.